File: Alkitab/src/main/java/yuku/alkitab/base/storage/OldVerseTextDecoder.java
package yuku.alkitab.base.storage;

import yuku.alkitab.io.Utf8Decoder;

import java.util.ArrayList;

public class OldVerseTextDecoder {
	public static class Ascii implements VerseTextDecoder {
		private void lowercase(byte[] ba) {
			for (int i = 0, blen = ba.length; i < blen; i++) {
				byte b = ba[i];
				if (b <= (byte)'Z' && b >= (byte)'A') {
					ba[i] |= 0x20;
				}
			}
		}
		
		@Override public String[] separateIntoVerses(byte[] ba, boolean lowercased) {
			ArrayList<String> versesBuf = new ArrayList<>(60);
			char[] verseBuf = new char[4000];
			int i = 0;
			
			if (lowercased) {
				lowercase(ba);
			}
			
			//# WARNING: This will work only if all bytes are less than 0x80. 
			for (byte c : ba) {
				if (c == (byte) 0x0a) {
					String single = new String(verseBuf, 0, i);
					versesBuf.add(single);
					i = 0;
				} else {
					verseBuf[i++] = (char) c;
				}
			}
			
			return versesBuf.toArray(new String[versesBuf.size()]);
		}

		@SuppressWarnings("deprecation") @Override public String makeIntoSingleString(byte[] ba, boolean lowercased) {
			if (lowercased) {
				lowercase(ba);
			}

			return new String(ba, 0);
		}
	}
	
	public static class Utf8 implements VerseTextDecoder {
		@Override public String[] separateIntoVerses(byte[] ba, boolean lowercased) {
			ArrayList<String> versesBuf = new ArrayList<>(60);
			
			int len = ba.length;
			int from = 0;
			for (int pos = 0; pos < len; pos++) {
				byte c = ba[pos];
				if (c == (byte)0x0a) {
					String single;
					if (lowercased) {
						single = Utf8Decoder.toStringLowerCase(ba, from, pos - from);
					} else {
						single = Utf8Decoder.toString(ba, from, pos - from);
					}
					versesBuf.add(single);
					from = pos + 1;
				}
			}
			
			return versesBuf.toArray(new String[versesBuf.size()]);
		}

		@Override public String makeIntoSingleString(byte[] ba, boolean lowercased) {
			String res;
			if (lowercased) {
				res = Utf8Decoder.toStringLowerCase(ba);
			} else {
				res = Utf8Decoder.toString(ba);
			}
			return res;
		}
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/storage/OldVerseTextDecoder.kt
package yuku.alkitab.base.storage

import yuku.alkitab.io.Utf8Decoder

object OldVerseTextDecoder {
    class Ascii : VerseTextDecoder {
        private fun lowercase(ba: ByteArray) {
            for (i in ba.indices) {
                if (ba[i] in 'A'.code..'Z'.code) {
                    ba[i] = (ba[i].toInt() or 0x20).toByte()
                }
            }
        }

        override fun separateIntoVerses(ba: ByteArray, lowercased: Boolean): Array<String> {
            val versesBuf = ArrayList<String>(60)
            val verseBuf = CharArray(4000)
            var i = 0

            if (lowercased) {
                lowercase(ba)
            }

            //# WARNING: This will work only if all bytes are less than 0x80.
            for (c in ba) {
                if (c == 0x0a.toByte()) {
                    versesBuf.add(String(verseBuf, 0, i))
                    i = 0
                } else {
                    verseBuf[i++] = c.toInt().toChar()
                }
            }

            return versesBuf.toTypedArray()
        }

        override fun makeIntoSingleString(ba: ByteArray, lowercased: Boolean): String {
            if (lowercased) {
                lowercase(ba)
            }

            return String(ba, Charsets.ISO_8859_1)
        }
    }

    class Utf8 : VerseTextDecoder {
        override fun separateIntoVerses(ba: ByteArray, lowercased: Boolean): Array<String> {
            val versesBuf = ArrayList<String>(60)

            var from = 0
            for (pos in ba.indices) {
                if (ba[pos] == 0x0a.toByte()) {
                    val single = if (lowercased) {
                        Utf8Decoder.toStringLowerCase(ba, from, pos - from)
                    } else {
                        Utf8Decoder.toString(ba, from, pos - from)
                    }
                    versesBuf.add(single)
                    from = pos + 1
                }
            }

            return versesBuf.toTypedArray()
        }

        override fun makeIntoSingleString(ba: ByteArray, lowercased: Boolean): String =
            if (lowercased) Utf8Decoder.toStringLowerCase(ba) else Utf8Decoder.toString(ba)
    }
}
